s3/client: add tagging ops

tchaikov opened the pull request:
with tagging ops, we will be able to attach kv pairs to an object. this will allow us to mark sstable components with taggings, and filter them based on them.

* test/pylib/minio_server.py: enable anonymous user to perform more actions. because the tagging related ops are not enabled by "mc anonymous set public", we have to enable them using "set-json" subcommand.
* utils/s3/client: add methods to manipulate taggings.
* test/boost/s3_test: add a simple test accordingly.

tchaikov commented:
@xemul hi Pavel, could you help review this change as well?

tchaikov commented:
this change might conflict with #14485

nyh commented on test/boost/s3_test.cc:
```diff
@@ -205,3 +205,28 @@ SEASTAR_THREAD_TEST_CASE(test_client_readable_file) {
+        auto tagset = client->get_object_tagging(name).get0();
+        BOOST_CHECK(tagset.empty());
+    }
+    client->delete_object(name).get();
```
This test uses a non-unique name for the test object, and although it tries to delete the object at the end of the test, if there is an exception in the middle of the test we won't reach this deletion. I think we should try harder. An easy solution is to create a RAII object which invents a unique object name, and also deletes the S3 object when the RAII object is destroyed.

tchaikov replied:
please note, `BOOST_CHECK()` allows a test case to proceed even if one of the tests fails. but yeah, these methods under test can throw. i will add another commit to address your comment using RAII. as other tests in this source file have the same problem.

nyh commented on test/pylib/minio_server.py:
```diff
@@ -52,6 +53,41 @@ def log_to_file(self, str):
 
+    def _anonymous_public_policy(self):
+        # the public policy does not allow accessing the tagging,
+        # so let's add them manually
```
The structure below has a bunch of weird details that I'm guessing you copied from somewhere. Might it be possible and perhaps better to write code that copies the existing policy file and *add* the Tagging actions instead of creating the whole thing from scratch?

I'm worried that a year from now, this hack will stop working because minio's "policy file" will have changed and will need additional things that you didn't copy.

But on the other hand, it's not really a big problem, if you continue to maintain this script.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2199/

nyh commented on utils/s3/client.cc:
```diff
@@ -223,6 +232,64 @@ future<client::stats> client::get_object_stats(sstring object_name) {
+            s3l.warn("cannnot parse tagging response: {}", e.what());
+            throw std::runtime_error("cannot parse tagging response");
+        }
+        auto tagset_node = doc->first_node("Tagging")->first_node("TagSet");
```
I'm not familiar with this "rapidxml" library, does first_node() throw an exception if there is no such node? Will the exception be something reasonable that the user/debugger/etc. be able to understand?

tchaikov replied:
good point. will early return in this case.

tchaikov replied:
see also https://rapidxml.sourceforge.net/manual.html#classrapidxml_1_1xml__node_7823e36687669e59c2afdf66334ef35a_17823e36687669e59c2afdf66334ef35a and https://github.com/discord/rapidxml/blob/2ae4b2888165a393dfb6382168825fddf00c27b9/rapidxml.hpp#L936-L946

nyh commented on utils/s3/client.cc:
```diff
@@ -223,6 +232,64 @@ future<client::stats> client::get_object_stats(sstring object_name) {
+            throw std::runtime_error("cannot parse tagging response");
+        }
+        auto tagset_node = doc->first_node("Tagging")->first_node("TagSet");
+        for (auto tag_node = tagset_node->first_node("Tag"); tag_node; tag_node = tag_node->next_sibling()) {
```
What happens here if there are zero tags? Above I assumed that first_node() throws if there is no such tag (otherwise, the code above could cause a null pointer dereference!) but here it seems it is a null pointer. So what's going on?

tchaikov replied:
it returns a `nullptr`. the newly added test actually verify this behavior.

nyh reviewed (commented):
Looks reasonable to me (I wrote some questions that may be based on my ignorance on the rapidxml library), but it's better that somebody more familiar with this code review it as well.

xemul commented on utils/s3/client.cc:
```diff
@@ -223,6 +232,64 @@ future<client::stats> client::get_object_stats(sstring object_name) {
+
+future<> client::delete_object_tagging(sstring object_name) {
+    // see https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjectTagging.html
+    auto req = http::request::make("DELETE", _host, fmt::format("{}?tagging", object_name));
```
Use `req.query_parameters["tagging"] = ""` instead

tchaikov replied:
ack.

xemul commented on utils/s3/client.cc:
```diff
@@ -223,6 +232,64 @@ future<client::stats> client::get_object_stats(sstring object_name) {
+
+future<> client::put_object_tagging(sstring object_name, client::TagSet tagging) {
+    // see https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObjectTagging.html
+    auto req = http::request::make("PUT", _host, fmt::format("{}?tagging", object_name));
```
Use req.query_parameters["tagging"] = "" instead

tchaikov replied:
sure.

xemul commented on utils/s3/client.cc:
```diff
@@ -223,6 +232,64 @@ future<client::stats> client::get_object_stats(sstring object_name) {
 
+future<client::TagSet> client::get_object_tagging(sstring object_name) {
+    // see https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObjectTagging.html
+    auto req = http::request::make("GET", _host, fmt::format("{}?tagging", object_name));
```
Use `req.query_parameters["tagging"]` = "" instead

tchaikov replied:
got it.

xemul commented on utils/s3/client.cc:
```diff
@@ -223,6 +232,64 @@ future<client::stats> client::get_object_stats(sstring object_name) {
+        auto& tagset = tagging;
+        auto input = std::move(in);
+        auto body = co_await util::read_entire_stream_contiguous(input);
+        auto doc = std::make_unique<rapidxml::xml_document<>>();
```
Please, put parsing into a dedicated helper function next to `parse_multipart_upload_id` and `parse_multipart_copy_upload_etag` so that we could collect all XML-bodies parsing in one place some day

tchaikov replied:
> Please, put parsing into a dedicated helper function

i put the parsing into a helper.

> next to parse_multipart_upload_id and parse_multipart_copy_upload_etag so that we could collect all XML-bodies parsing in one place some day

why do we want to put them together?  i am not sure it's better to group the helpers together is an improvement though. for two reasons:

1. `parse_multipart_upload_id()` and `parse_multipart_copy_upload_etag()` are helpers for the methods of `client::upload_sink_base`. this class is in turn used by `client::make_upload_sink()` and `client::make_upload_jumbo_sink()`. i think that's why they are grouped together. for the same reason, `client::readable_file` and `client::make_readable_file()` are co-located. this helps with the readability. as the declarations and definitions designed as a part of the similar functionality are grouped together.  no need to jump back and forth when reading the source code. he or she can just read through them. the same applies to `parse_http_last_modified_time()` and `client::get_object_stats()`. if we group the parsers and printers together, it would hurt the readability and maintainability.
2. from a more practical perspective, we don't need to have a forward declarations of the helper or place the definitions of helpers before their consumer. please note, i intentionally put the definition of the tagging related function in the same order of their declarations. and the methods' declarations are ordered from the minimal stats to an object's attributes, and to its data.

xemul commented on utils/s3/client.hh:
```diff
@@ -51,6 +51,15 @@ public:
         std::time_t last_modified;
     };
     future<stats> get_object_stats(sstring object_name);
+    struct Tag {
```
Why `Tag` but not `tag`?
I would also put it at namespace scope next to the `s3::range` struct in the same header

tchaikov replied:
i wanted to match the names in its API definition, see https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketTagging.html. but `tag` is fine by me, if we want to use lower-cased names for the data-only structs.

xemul commented on utils/s3/client.cc:
```diff
@@ -223,6 +232,64 @@ future<client::stats> client::get_object_stats(sstring object_name) {
+    auto req = http::request::make("PUT", _host, fmt::format("{}?tagging", object_name));
+    s3l.warn("PUT {} tagging", object_name);
+    fmt::memory_buffer body;
+    fmt::format_to(fmt::appender(body), "<Tagging><TagSet>{}</TagSet></Tagging>", fmt::join(tagging, ""));
```
Please, put both -- size calculation and body-writer -- as a local helper functions next to `prepare_multipart_upload_parts` and `dump_multipart_upload_parts`

tchaikov replied:
i extracted the dump logic into a helper. for the same reason explained above i didn't put it next to this helper next to `prepare_multipart_upload_parts()` and `dump_multipart_upload_parts()`.

tchaikov commented:
v2

1. add comment to explain how the anonymous policy is created.
2. bail out if the expected node is missing when parsing the reply from s3 server
3. extract the parser and printer for response and request into separate helpers
4. user lower-case names for `Tag` and `TagSet`, and move them into `s3` namespace
5. use RAII for clean ups in the test
6. use different names for the objects created in different test cases.

tchaikov commented:
@nyh @xemul hi Nadav and Pavel, thank you for your reviews. could you take another look?

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2223/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2227/

xemul commented on test/boost/s3_test.cc:
```diff
@@ -61,7 +61,7 @@ s3::endpoint_config_ptr make_minio_config() {
 
 SEASTAR_THREAD_TEST_CASE(test_client_put_get_object) {
-    const sstring name(fmt::format("/{}/testobject-{}", tests::getenv_safe("S3_PUBLIC_BUCKET_FOR_TEST"), ::getpid()));
+    const sstring name(fmt::format("/{}/testobject-{}", tests::getenv_safe("S3_PUBLIC_BUCKET_FOR_TEST"), get_name()));
```
> instead of using pid of current process, use the test's name for naming object used in test. less likely to cause conflicts if some test fails to perform the cleanup as expected.

Do test names differ for same test case running in different modes (dev/release/debug)?

tchaikov replied:
no. the test names do not encode the test mode, AFAICT. 

and i see your concerns -- `test.py` can drive different builds and they share the same minio server. so i still need to encode an unique id in the object name. i will amend this commit to take this into consideration. probably will have a tiny helper for this purpose.

tchaikov replied:
after a second thought, I think the “use the same name” issue is a false one, as the way how we run the test is immune to this problem. I will drop this commit from this PR.

xemul commented:
@tchaikov , only the 1st patch meets the PR title, 3 others are on their own, aren't they?

tchaikov commented:
> @tchaikov , only the 1st patch meets the PR title, 3 others are on their own, aren't they?

sorry. i should have update the title. now it should be better.

tchaikov commented:
v3:

- use a different object name in "test_client_put_get_tagging" than the one used by "test_client_put_get_object"
- drop the commit to name object after the name of test. as we are not likely the name conflict if we encode the object's key with the pid and the test specific name.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2259/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2272/

tchaikov commented:
@xemul mind taking another look? i should have addressed your comments.

xemul commented on utils/s3/client.cc:
```diff
@@ -223,6 +234,92 @@ future<client::stats> client::get_object_stats(sstring object_name) {
+        auto input = std::move(in);
+        auto body = co_await util::read_entire_stream_contiguous(input);
+        retval = parse_tagging(body);
+    }, http::reply::status_type::ok);
```
nit: `status_type::ok` is default arg. Can be a followup pr

tchaikov replied:
will drop it tomorrow.

xemul commented on utils/s3/client.cc:
```diff
@@ -28,6 +30,15 @@
+struct fmt::formatter<s3::tag> {
+    constexpr auto parse(format_parse_context& ctx) { return ctx.begin(); }
+    auto format(const s3::tag& tag, fmt::format_context& ctx) const {
+        return fmt::format_to(ctx.out(), "<Tag><Key>{}</Key><Value>{}</Value></Tag>",
```
I heavily doubt that making formatter of whatever type produce XML string is a good idea

tchaikov replied:
hmm, kind of. the same idea came to my mind also. IMHO, to concatenate strings and to print using fmtlib have their pros and cons.. but since it's part of an implementation, and is not horribly ugly, so i went ahead.

xemul commented:
> > @tchaikov , only the 1st patch meets the PR title, 3 others are on their own, aren't they?
> 
> sorry. i should have updated title. now it should be better.

Please, split the PR. Enhancing test cleanup has nothing to do with adding new functionality regarding tagging and it makes git archaeology in the future more wtf-prone

tchaikov commented:
> > > @tchaikov , only the 1st patch meets the PR title, 3 others are on their own, aren't they?
> > 
> > 
> > sorry. i should have updated title. now it should be better.
> 
> Please, split the PR. Enhancing test cleanup has nothing to do with adding new functionality regarding tagging and it makes git archaeology in the future more wtf-prone

i see. i must have misunderstood your last comment. will do tomorrow.

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request

tchaikov commented:
v3:

1. drop the cleanup changes 
2. drop the unnecessary `http::reply::status_type::ok` parameter

@xemul hi Pavel, thank you for your review. the latest revision should have addressed your comments. could you help review it again?

tchaikov commented:
v4

- rebase against master HEAD to resolve the conflict in `test/pylib/minio_server.py`.

tchaikov commented on utils/s3/client.cc:
```diff
@@ -223,6 +234,92 @@ future<client::stats> client::get_object_stats(sstring object_name) {
+        }
+    }
+    return node;
+}
```
note to myself, we can reuse this helper in other places in this source file where we want to reference an element with XPath like 'foo/bar'. despite that the API document might claim that there will be one or more bar under `foo`, we should handle the cases where `foo` or `foo/bar` does not exist, and error out if it really happens.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2289/

tchaikov commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2289/

it's a known issue, tracked by #10527.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2291/

tchaikov commented:
@scylladb/scylla-maint  hi maintainers, could you please queue this change?

tchaikov commented:
> v3:
> 
>     1. drop the cleanup changes
> 

the cleanups were extracted into #14765
